File: aquacrop/CropYield.py

#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np

from hm import file_handling

import logging
logger = logging.getLogger(__name__)

class CropYield(object):
    def __init__(self, CropYield_variable):
        self.var = CropYield_variable

    def initial(self):
        arr_zeros = np.zeros((self.var.nFarm, self.var.nCrop, self.var.nCell))
        self.var.CropMature = np.copy(arr_zeros).astype(bool)
        self.var.Y = np.copy(arr_zeros)

    def reset_initial_conditions(self):
        self.var.CropMature[self.var.GrowingSeasonDayOne] = False
        
    def dynamic(self):
        """Function to calculate crop yield"""
        if np.any(self.var.GrowingSeasonDayOne):
            self.reset_initial_conditions()

        cond1 = self.var.GrowingSeasonIndex
        self.var.Y[cond1] = ((self.var.B / 100) * self.var.HIadj)[cond1]

        # print('Y    :',self.var.Y[...,500])
        # print('B    :',self.var.B[...,500])
        # print('HIadj:',self.var.HIadj[...,500])
        
        cond11 = (cond1 & (((self.var.CalendarType == 1) & ((self.var.DAP - self.var.DelayedCDs) >= self.var.Maturity)) | ((self.var.CalendarType == 2) & ((self.var.GDDcum - self.var.DelayedGDDs) >= self.var.Maturity))))
        self.var.CropMature[cond11] = True
        self.var.Y[np.logical_not(self.var.GrowingSeasonIndex)] = 0

File: aquacrop/native/crop_yield.f90
module crop_yield
  use types
  implicit none

contains

  subroutine update_crop_yield( &
       yield, &
       crop_mature, &
       maturity, &
       b, &
       hi_adj, &
       gdd_cum, &
       growing_season, &
       growing_season_day1, &
       calendar_type, &
       dap, &       
       delayed_cds, &
       delayed_gdds &
       )

    real(real64), intent(inout) :: yield
    integer(int32), intent(inout) :: crop_mature
    real(real64), intent(in) :: maturity
    real(real64), intent(in) :: b
    real(real64), intent(in) :: hi_adj
    real(real64), intent(in) :: gdd_cum
    integer(int32), intent(in) :: growing_season
    integer(int32), intent(in) :: growing_season_day1
    integer(int32), intent(in) :: calendar_type
    integer(int32), intent(in) :: dap
    integer(int32), intent(in) :: delayed_cds
    real(real64), intent(in) :: delayed_gdds

    if ( growing_season_day1 == 1 ) then
       crop_mature = 0       
    end if

    if ( growing_season == 1 ) then
       
       yield = (b / 100.) * hi_adj
       if ( calendar_type == 1 ) then
          if ( (dap - delayed_cds) >= maturity ) then
             crop_mature = 1
          else
             crop_mature = 0
          end if
          
       else if ( calendar_type == 2 ) then
          if ( (gdd_cum - delayed_gdds) >= maturity ) then
             crop_mature = 1
          else
             crop_mature = 0
          end if
       end if       
    else
       yield = 0
    end if
    
  end subroutine update_crop_yield
  
end module crop_yield


    
